adr/keycloak-adr.md

```diff
@@ -0,0 +1,236 @@
+   - **High Performance**: Ensuring high throughput and low latency, ideal for real-time applications.
+   - **Fault Tolerance**: Through clustering, NATS achieves high availability, safeguarding against node failures.
+   - **Scalability**: Capable of handling a growing number of clients and messages, efficiently distributing messages across subscribers.
+   - **Security**: Supports TLS/SSL for secure communication between clients and servers.
+
+3. **Service for API Updates**: A dedicated service will subscribe to the NATS topic to process messages related to user deactivation or deletion. This service will then interact with the APIs of GitLab, Mattermost, SonarQube, and other integrated services to revoke the user's access. This approach ensures a seamless and automated deprovisioning process across all services, enhancing our security posture.
+
+This design not only automates the user deprovisioning process but also establishes an event-driven architecture that minimizes tight coupling between services. By leveraging NATS within a Kubernetes environment, we ensure that our architecture is both scalable and resilient, capable of adapting to the evolving needs of our infrastructure and services.
+
+```mermaid
+graph TD
+    A[Security Enhancement with Event-Driven Architecture] --> B[Decoupling Keycloak and GitLab]
+    A --> C[Asynchronous Communication & Queuing]
+    A --> D[Improved Fault Tolerance & Reliability]
+    A --> E[Guaranteed Delivery & Processing]
+
+    B --> B1[Eliminates Direct Dependencies]
+    B --> B2[Isolates System Components]
+
+    C --> C1[No Immediate Need for Receiver Availability]
+    C --> C2[Messages Persisted Until Processed]
+
+    D --> D1[Service Downtime Does Not Result in Data Loss]
+    D --> D2[Independent Recovery Mechanisms]
+
+    E --> E1[Retries Until Successful Processing]
+    E --> E2[Ensures User Deletion Events Are Handled]
+
+    B1 --> F[Reduced Security Risks]
+    C2 --> G[No Loss of Critical Security Events]
+    D2 --> H[Resilience Against System Failures]
+    E2 --> I[Compliance with Security Policies]
+
+    F --> J[Enhanced Overall Security Posture]
+    G --> J
+    H --> J
+    I --> J
+```
+The provided sequence diagram illustrates a scalable and decoupled architecture for handling user deprovisioning, leveraging NATS for centralized publish/subscribe messaging. This pattern ensures that when a user is deleted or deactivated in Keycloak, the event is published to NATS, from where it is distributed to various service listeners, including a GitLab Service Listener and an Audit Listener. The GitLab Service Listener is responsible for revoking the user's access in GitLab, while the Audit Listener updates the Admin UI to display the user deletion event. This architecture can be easily extended to include additional services by adding more listeners that subscribe to the relevant NATS topics for user status changes.
+```mermaid
+sequenceDiagram
```

Comment:
I think a lot of this information can be saved for a later design doc - would like to see more Pro/Con / tradeoff analysis of general approaches.